fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/identifiers/FedoraID.java

```diff
@@ -0,0 +1,197 @@
+    public boolean isHashUri() {
+        return hashUri != null;
+    }
+
+    /**
+     * Get the hash uri.
+     * @return the hash uri from the id or null if none.
+     */
+    public String getHashUri() {
+        return hashUri;
+    }
+
+    /**
+     * Return the ID of the base resource for this request.
+     * @return the shorten id.
+     */
+    public String getResourceId() {
+        return id;
+    }
+
+    /**
+     * Return the original full ID.
+     * @return the id.
+     */
+    public String getFullId() {
+        return fullId;
+    }
+
+    /**
+     * Return the Memento datetime as Instant.
+     * @return The datetime or null if not a memento.
+     */
+    public Instant getMementoInstant() {
+        return mementoDatetime;
+    }
+
+    /**
+     * Return the Memento datetime string.
+     * @return The yyyymmddhhiiss memento datetime or null if not a Memento.
+     */
+    public String getMementoString() {
+        return mementoDatetime_str;
+    }
+
+    /**
+     * Process the original ID into its parts without using a regular expression.
+     */
+    private void processIdentifier() {
+        // Regex pattern which decomposes a http resource uri into components
+        // The first group determines if it is an fcr:metadata non-rdf source.
+        // The second group determines if the path is for a memento or timemap.
+        // The third group allows for a memento identifier.
+        // The fourth group for allows ACL.
+        // The fifth group allows for any hashed suffixes.
+        // ".*?(/" + FCR_METADATA + ")?(/" + FCR_VERSIONS + "(/\\d{14})?)?(/" + FCR_ACL + ")?(\\#\\S+)?$");
+        String processID = this.fullId;
+        if (processID.equals(FEDORA_ID_PREFIX)) {
+            this.isRepositoryRoot = true;
+            this.id = this.fullId;
+            // Root has no other possible endpoints, so short circuit out.
+            return;
+        }
+        if (processID.contains("#")) {
+            this.hashUri = processID.split("#")[1];
+            processID = processID.split("#")[0];
+        }
+        if (processID.contains(FCR_ACL)) {
+            this.isAcl = true;
+            processID = processID.split(FCR_ACL)[0];
+        }
+        if (processID.contains(FCR_VERSIONS)) {
+            if (processID.split(FCR_VERSIONS).length == 2) {
+                final String afterVersion = processID.split(FCR_VERSIONS)[1];
+                if (afterVersion.matches("/\\d{14}")) {
+                    this.isMemento = true;
+                    this.mementoDatetime_str = afterVersion.substring(1);
+                    this.mementoDatetime = Instant.from(MEMENTO_LABEL_FORMATTER.parse(this.mementoDatetime_str));
+                } else {
+                    this.isTimemap = true;
+                }
+            } else {
+                this.isTimemap = true;
+            }
+            processID = processID.split(FCR_VERSIONS)[0];
+        }
+        if (processID.contains(FCR_METADATA)) {
```

Comment:
Similar question here as the acl question above. At this point, there shouldn't be anything after `fcr:metadata`, right? If there is, should it be ignored or rejected?

Reply:
Same question here, what do we want? Fedora 5.1 does this:
```
> curl -ufedoraAdmin:fedoraAdmin -i http://localhost:8080/rest/binary1/fcr:metadata/test-garbage
HTTP/1.1 400 Bad Request
Date: Mon, 23 Mar 2020 15:57:55 GMT
Set-Cookie: JSESSIONID=1nsw8b5a9u9ok3kh3dzl5nnmn;Path=/
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Set-Cookie: rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Sun, 22-Mar-2020 15:57:55 GMT
Content-Type: text/plain;charset=utf-8
Content-Length: 53
Server: Jetty(9.3.25.v20180904)

Error converting "fcr:metadata" from String to a Name
```

Reply:
Resolved with 608e69dd845edb56d36941606048ebcfa493634f